tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -0,0 +1,331 @@
+    ident: NodeAndRange
+    resolved_to: NodeAndRange
+
+
+@dataclass
+class FileInfo:
+    uri: str
+    context: chapel.core.Context
+    segments: List[ResolvedPair] = field(default_factory=list)
+    siblings: chapel.SiblingMap = field(init=False)
+
+    def __post_init__(self):
+        self.rebuild_segments()
+
+    def parse_file(self) -> List[chapel.core.AstNode]:
+        """
+        Given a file URI, return the ASTs making up that file. Advances
+        the context if one already exists to make sure an updated result
+        is returned.
+        """
+
+        return self.context.parse(self.uri[len("file://") :])
+
+    def get_asts(self) -> List[chapel.core.AstNode]:
+        with self.context.track_errors() as _:
+            return self.parse_file()
+
+    def rebuild_segments(self):
+        asts = self.get_asts()
+        # get ids
+        self.segments = [
+            ResolvedPair(NodeAndRange(node), NodeAndRange(to))
+            for node, _ in chapel.each_matching(asts, chapel.core.Identifier)
+            if (to := node.to_node())
+        ]
+        self.segments.sort(key=lambda s: s.ident.rng.start)
+        self.siblings = chapel.SiblingMap(self.get_asts())
+
+
+def run_lsp():
+    """
+    Start a language server on the standard input/output, and use it to
+    report linter warnings as LSP diagnostics.
+    """
+
+    server = LanguageServer("chplcheck", "v0.1")
+
+    contexts: Dict[str, FileInfo] = {}
+
+    def get_context(uri: str, do_update: bool = False) -> Tuple[FileInfo, List[Any]]:
+        """
+        The LSP driver maintains one Chapel context per-file. This is to avoid
+        having to reset all files' text etc. when a single file is updated.
+        There may be a more principled approach we can take in the future.
+
+        This function returns an _update_ context, which is effectively a context
+        in which we can save / make use of updated file text. If there wasn't
+        a context for a URI, a brand new context will do. For existing contexts,
+        an older version of the file's text is probably stored, so advance
+        the context to next revision, invalidating that cache.
+
+        Thus, this method is effectively allocate-or-advance-context.
+        """
+
+        errors = []
+
+        if uri in contexts:
+            file_info = contexts[uri]
+            if do_update:
+                file_info.context.advance_to_next_revision(False)
+                with file_info.context.track_errors() as errors:
+                    file_info.rebuild_segments()
+        else:
+            context = chapel.core.Context()
+            with context.track_errors() as errors:
+                file_info = FileInfo(uri, context)
+            contexts[uri] = file_info
+
+        return (file_info, errors)
+
+    def build_diagnostics(uri):
+        """
+        Parse a file at a particular URI, run the linter rules on the resulting
+        ASTs, and return them as LSP diagnostics.
```

Comment:
Copy paste-o: linter rules? I think this method is effectively parse-and-print-syntax-errors now.